barretenberg/cpp/src/barretenberg/dsl/acir_format/acir_to_constraint_buf.hpp

```diff
@@ -4,18 +4,43 @@
 
 namespace acir_format {
 
-AcirFormat circuit_buf_to_acir_format(std::vector<uint8_t> const& buf, uint32_t honk_recursion);
+/**
+ * @brief Deserialize `buf` either based on the first byte interpreted as a
+          Noir serialization format byte, or falling back to `bincode` if
+          the format cannot be recognized. Currently only `msgpack` format
+          is expected, or the legacy `bincode` format.
+ * @note Due to the lack of exception handling available to us in Wasm we can't
+ *       try `bincode` format and if it fails try `msgpack`; instead we have to
+ *       make a decision and commit to it.
+ */
+template <typename T>
+T deserialize_any_format(std::vector<uint8_t> const& buf,
+                         std::function<T(msgpack::object const&)> decode_msgpack,
+                         std::function<T(std::vector<uint8_t>)> decode_binpack);
```

Comment:
If you only use this one inside the cpp then remove it from the hpp.

Otherwise, if you use it from other places, it should actually not work. As it is templated, it will not work unless the implementation is in the hpp as well.

But ideally don't move the implementation to the hpp, but just keep everything in the cpp. Check the same for other functions as well. If they are only used in a cpp, they should be in that cpp (in an anonymous namespace ideally).

Reply:
Thanks, yeah, that's what I thought too. Initially I added some function to the cpp and it complained that it can't find it; I moved it to the hpp and it was fine. Not what I expected, but I had other problems. I'll try to move this back.

Reply:
Done.